### PR TITLE
Hare xdr

### PR DESCRIPTION
**Pull request by gavraz**
* Replaced protobuf with XDR for hare protocol

### CODE REVIEW

File: hare/builder.go [hunk 1]
```diff
@@ -1,19 +1,49 @@
...
-	"github.com/gogo/protobuf/proto"
-	"github.com/spacemeshos/go-spacemesh/hare/pb"
+	"bytes"
+	"github.com/nullstyle/go-xdr/xdr3"
 	"github.com/spacemeshos/go-spacemesh/log"
 )
 
-// Used to build proto messages
+// top Message of the protocol
+type XDRMessage struct {
```
Comment (antonlerner) [inline hare/builder.go]: Why not  name it just message? the encoding method is irrelevant

File: hare/algorithm.go [hunk 2]
```diff
@@ -255,39 +256,44 @@ func (proc *ConsensusProcess) onEarlyMessage(m *Msg) {
 		return
 	}
 
-	if m.HareMessage == nil {
-		proc.Error("onEarlyMessage called with nil message")
+	if m.XDRMessage == nil {
```
Comment (antonlerner) [inline hare/algorithm.go]: rename to hare message

File: hare/algorithm.go [hunk 3]
```diff
@@ -255,39 +256,44 @@ func (proc *ConsensusProcess) onEarlyMessage(m *Msg) {
...
 	if !proc.validator.ContextuallyValidateMessage(m, proc.k) {
 		if !proc.validator.ContextuallyValidateMessage(m, proc.k+1) {
-			// TODO: should return error from message validation to indicate what failed, should retry only for contextual failure
+			// TODO: should return error from Message validation to indicate what failed, should retry only for contextual failure
 			proc.Warning("Message of type %v is not valid for either round, pubkey %v", mType, m.PubKey)
 			return
-		} else { // a valid early message, keep it for later
-			proc.Info("Early message of type %v detected. Keeping message, pubkey %v", mType, m.PubKey)
+		} else { // a valid early Message, keep it for later
+			proc.Info("Early Message of type %v detected. Keeping Message, pubkey %v", mType, m.PubKey)
```
Comment (antonlerner) [inline hare/algorithm.go]: use bytes to hex on pubkey, show only 5 characters from pubkey

File: hare/algorithm.go [hunk 4]
```diff
@@ -255,39 +256,44 @@ func (proc *ConsensusProcess) onEarlyMessage(m *Msg) {
...
 
 	proc.pending[pub.String()] = m
 }
 
 func (proc *ConsensusProcess) handleMessage(m *Msg) {
-	// Note: instanceId is already verified by the broker
+	// Note: InstanceId is already verified by the broker
 
-	proc.Debug("Received message %v", m)
+	proc.Debug("Received Message %v", m)
```
Comment (antonlerner) [inline hare/algorithm.go]: use a pretty printer of the message

File: hare/algorithm.go [hunk 5]
```diff
@@ -255,39 +256,44 @@ func (proc *ConsensusProcess) onEarlyMessage(m *Msg) {
...
 	}
 
-	if m.HareMessage == nil {
-		proc.Error("onEarlyMessage called with nil message")
+	if m.XDRMessage == nil {
+		proc.Error("onEarlyMessage called with nil Message")
+		return
+	}
+
+	if m.Message == nil {
```
Comment (antonlerner) [inline hare/algorithm.go]: should you process with concensus process after receiving a null message

File: hare/algorithm.go [hunk 6]
```diff
@@ -331,21 +337,22 @@ func (proc *ConsensusProcess) sendMessage(msg *Msg) {
 		return
 	}
 
-	data, err := proto.Marshal(msg.HareMessage)
+	var w bytes.Buffer
```
Comment (antonlerner) [inline hare/algorithm.go]: encapsulate in msgtobytes method

File: hare/algorithm.go [hunk 7]
```diff
@@ -567,8 +574,8 @@ func (proc *ConsensusProcess) endOfRound3() {
 		return
 	}
 
-	// commit & send notification message
-	proc.Debug("end of round 3: committing on %v and sending notification message", s)
+	// commit & send notification Message
+	proc.Debug("end of round 3: committing on %v and sending notification Message", s)
```
Comment (antonlerner) [inline hare/algorithm.go]: pretty print message

File: hare/broker.go [hunk 8]
```diff
@@ -93,24 +93,25 @@ func (b *Broker) eventLoop() {
 				continue
 			}
 
-			hareMsg := &pb.HareMessage{}
-			err := proto.Unmarshal(msg.Bytes(), hareMsg)
+			rdr := bytes.NewReader(msg.Bytes())
```
Comment (antonlerner) [inline hare/broker.go]: encapsulate in bytes to message func

File: hare/broker.go [hunk 9]
```diff
@@ -139,12 +140,12 @@ func (b *Broker) eventLoop() {
...
 				c <- iMsg
 				continue
-			} else { // message arrived before registration
+			} else { // Message arrived before registration
 				futureMsg = true
 			}
 
 			if futureMsg {
-				b.Info("Broker identified future message")
+				b.Info("Broker identified future Message")
```
Comment (antonlerner) [inline hare/broker.go]: I'd add more info to this message or remove it as it is not very informative.

File: hare/messagevalidation.go [hunk 10]
```diff
@@ -43,11 +42,11 @@ func (ev *eligibilityValidator) validateRole(m *Msg) bool {
 	return true
 }
 
-// Validates eligibility and signature of the provided message
+// Validates eligibility and signature of the provided Message
 func (ev *eligibilityValidator) Validate(m *Msg) bool {
 	// verify role
 	if !ev.validateRole(m) {
-		ev.Warning("Validate message failed: role is invalid")
+		ev.Warning("Validate Message failed: role is invalid")
```
Comment (antonlerner) [inline hare/messagevalidation.go]: this error is also not informative

File: hare/algorithm.go [hunk 11]
```diff
@@ -331,21 +337,22 @@ func (proc *ConsensusProcess) sendMessage(msg *Msg) {
...
 		return
 	}
 
-	proc.Info("Message of type %v sent", MessageType(msg.Message.Type).String())
+	proc.Info("Message of type %v sent", msg.Message.Type.String())
 }
 
 func (proc *ConsensusProcess) onRoundEnd() {
-	proc.With().Info("End of round", log.Int32("k", proc.k))
+	proc.With().Info("End of round", log.Int32("K", proc.k))
```
Comment (antonlerner) [inline hare/algorithm.go]: move to debug

File: hare/algorithm.go [hunk 12]
```diff
@@ -331,21 +337,22 @@ func (proc *ConsensusProcess) sendMessage(msg *Msg) {
...
 
-	if err := proc.network.Broadcast(protoName, data); err != nil {
-		proc.Error("Could not broadcast round message ", err.Error())
+	if err := proc.network.Broadcast(protoName, w.Bytes()); err != nil {
+		proc.Error("Could not broadcast round Message ", err.Error())
 		return
 	}
 
-	proc.Info("Message of type %v sent", MessageType(msg.Message.Type).String())
+	proc.Info("Message of type %v sent", msg.Message.Type.String())
```
Comment (antonlerner) [inline hare/algorithm.go]: move to debug

File: hare/algorithm.go [hunk 13]
```diff
@@ -298,9 +304,9 @@ func (proc *ConsensusProcess) handleMessage(m *Msg) {
 }
 
 func (proc *ConsensusProcess) processMsg(m *Msg) {
-	proc.Info("Processing message of type %v", MessageType(m.Message.Type).String())
+	proc.Info("Processing Message of type %v", m.Message.Type.String())
```
Comment (antonlerner) [inline hare/algorithm.go]: debug

**Review by antonlerner — CHANGES_REQUESTED** [review]
review comments

File: hare/builder.go (hunk 1 above)
Comment (gavraz) [reply hare/builder.go]: Fixed

File: hare/algorithm.go (hunk 2 above)
Comment (gavraz) [reply hare/algorithm.go]: Renamed to Message, not HareMessage

File: hare/algorithm.go (hunk 3 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed

File: hare/algorithm.go (hunk 4 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed

File: hare/algorithm.go (hunk 5 above)
Comment (gavraz) [reply hare/algorithm.go]: Shouldn't be possible, prefer an error

File: hare/algorithm.go (hunk 6 above)
Comment (gavraz) [reply hare/algorithm.go]: Fixed

File: hare/algorithm.go (hunk 7 above)
Comment (gavraz) [reply hare/algorithm.go]: Not a problem

File: hare/broker.go (hunk 8 above)
Comment (gavraz) [reply hare/broker.go]: fixed

File: hare/broker.go (hunk 9 above)
Comment (gavraz) [reply hare/broker.go]: Leftover from debugging. Removed

File: hare/messagevalidation.go (hunk 10 above)
Comment (gavraz) [reply hare/messagevalidation.go]: Fixed

File: hare/algorithm.go (hunk 11 above)
Comment (gavraz) [reply hare/algorithm.go]: ok

File: hare/algorithm.go (hunk 12 above)
Comment (gavraz) [reply hare/algorithm.go]: Ill keep it for now, not many msgs

File: hare/algorithm.go (hunk 13 above)
Comment (gavraz) [reply hare/algorithm.go]: ok

**Review by antonlerner — APPROVED** [review]

**Comment by zalmen** [comment]
@gavraz I would expect that this PR will also delete few pb files, why are we keeping them?

**Pull request merged**